### ISSUE TITLE
extended_glob cannot be disabled in zshrc

### ISSUE THREAD

**Issue by AlexWillisson** [ohmyzsh/ohmyzsh#10967]
This is an issue for me, I commented on the old issue in the hopes that it would reopen, but since it didn't I'm opening a new issue. If I `unsetopt extended_glob` manually, it works. If I put it as the last line in my `.zshrc`, it doesn't. I opened a new terminal in MacOS and copy/pasted everything from it here to show what I mean. I regularly use `^` for looking at git commits referenced relatively, I'm aware that `~` works too but I prefer `^`

```Last login: Wed Jun  1 11:50:21 on ttys008
awillisson@awillisson ~ » tail -5 .zshrc
export PATH=$PATH:$HOME/bin

unsetopt share_history
unsetopt EXTENDED_GLOB
awillisson@awillisson ~ » echo ^
<listing of files & directories in my home directory>
awillisson@awillisson ~ » unsetopt EXTENDED_GLOB
awillisson@awillisson ~ » echo ^
^
awillisson@awillisson ~ » 
```

I copy pasted `unsetopt EXTENDED_GLOB` to be 100% sure I didn't typo it in my `.zshrc`

_Originally posted by @AlexWillisson in https://github.com/ohmyzsh/ohmyzsh/issues/8588#issuecomment-1143787214_

**Comment by mcornella** [comment]
There's probably a function that is turning that on somewhere. To debug this you could start zsh with trace mode, have it output all the traces on a file, and exit the zsh session once you've confirmed that extended_glob was re-enabled (you can run `setopt | grep extendedglob` to check).

The command below will capture all the traces:
```sh
zsh -xv 2> >(tee ~/omz-debug.log &>/dev/null)
```

**Comment by mcornella** [comment]
Friendly ping @AlexWillisson

**Comment by TravisCarden** [comment]
@mcornella, I have the same issue. I captured the traces as you suggested. The resulting log was 9,640 lines, and I'm afraid of unwittingly disclosing sensitive information by providing the whole thing here, so here's a  grep:

```
$ tail -1 ~/.zshrc
unsetopt EXTENDED_GLOB

$ zsh -xv 2> >(tee ~/omz-debug.log &>/dev/null)

$ exit

$ ag -C3 EXTENDED_GLOB ~/omz-debug.log
9577-+compdef:169> shift
9578-+compdef:131> ((  0  ))
9579-
9580:unsetopt EXTENDED_GLOB
9581:+/Users/traviscarden/.zshrc:137> unsetopt EXTENDED_GLOB
9582-+omz_termsupport_precmd:1> [[ '' != true ]]
9583-+omz_termsupport_precmd:2> title '%15<..<%~%<<' '%n@%m:%~'
9584-+title:1> setopt localoptions nopromptsubst
```

I went so far as to recursively grep my entire home directory without finding anything unexpected:

```
$ ag --literal --word-regexp --unrestricted EXTENDED_GLOB ~/
/Users/traviscarden/omz-debug.log
9580:unsetopt EXTENDED_GLOB
9581:+/Users/traviscarden/.zshrc:137> unsetopt EXTENDED_GLOB

/Users/traviscarden/.zshrc
137:unsetopt EXTENDED_GLOB
```

I don't see anything promising there. 🤔 Let me know what else I can provide.

**Workaround** for anyone using iTerm as their terminal, set `unsetopt EXTENDED_GLOB` as [a custom profile command](https://apple.stackexchange.com/questions/115598/how-to-set-an-iterm2-startup-command) to have it run whenever you open a new iTerm session. It gets the job done.

**Comment by rwmitchell** [comment]
I don't use ag, was your search case insensitive ?

I used ```grep -i "extended.*glob"  **/(.*|*)(.)'``

and found it was being set in .p10k.zsh:

```
grep -i "extended.*glob" RF 
.p10k.zsh:  emulate -L zsh -o extended_glob
.p10k.zsh:  #      extended_glob option enabled.
```

however this doesn't appear to actually leak out into my shell as setopt doesn't show it being set.

**Comment by TravisCarden** [comment]
Ah no, @rwmitchell: my search was _not_ case insensitive. Here're the results with your grep command:

```
grep -i "extended.*glob" ~/omz-debug.log
+compinit:71> setopt extendedglob
+compinit:131> _comp_options=( bareglobqual extendedglob glob multibyte multifuncdef nullglob rcexpandparam unset NO_allexport NO_aliases NO_cshnullglob NO_cshjunkiequotes NO_errexit NO_errreturn NO_globassign NO_globsubst NO_histsubstpattern NO_ignorebraces NO_ignoreclosebraces NO_kshglob NO_ksharrays NO_kshtypeset NO_markdirs NO_octalzeroes NO_posixbuiltins NO_posixidentifiers NO_shwordsplit NO_shglob NO_warnnestedvar NO_warncreateglobal )
+compaudit:14> setopt extendedglob
+compdef:5> setopt extendedglob
+compdef:5> setopt extendedglob
  setopt extended_glob
+compdef:5> setopt extendedglob
+compdef:5> setopt extendedglob
unsetopt EXTENDED_GLOB
+/Users/traviscarden/.zshrc:137> unsetopt EXTENDED_GLOB
+omz_termsupport_preexec:4> setopt extended_glob
```

If I'm reading that right, `omz_termsupport_preexec` looks like the culprit. But I have no idea what it is. I can't find such a string or a filename anywhere. I'm sure it means more to you.

**Comment by mcornella** [comment]
`omz_termsupport_preexec` is the function that runs the auto-title functionality before a command is run (usually it sets the running command as the title).

It is run here:
https://github.com/ohmyzsh/ohmyzsh/blob/a7d910c3a61d8599f748a8ddae59ecdd9424022a/lib/termsupport.zsh#L58-L59

However, as you see it is preceded by `emulate -L zsh`, which sets [`setopt localoptions`](https://zsh.sourceforge.io/Doc/Release/Options.html#index-LOCALOPTIONS).

With this option, the `setopt` commands inside the function don't affect the environment outside the function.

```sh
unsetopt extendedglob

# this is an anonymous function
() {
  setopt | grep -q extendedglob || echo before: no
  emulate -L zsh
  setopt extendedglob
  setopt | grep -q extendedglob && echo function: yes
}

setopt | grep -q extendedglob || echo after: no
```

This will say

```
before: no
function: yes
after: no
```

The same is true with the construct `setopt localoptions <some-other-option>` BTW, it's not necessary to call `emulate -L zsh` to get this behavior.

TL;DR this means that this is not the culprit, and there must be something else after this command that we are not seeing.

Perhaps the issue is that `extendedglob` has not been enabled at the point where you ran exit when capturing the trace. Make sure that you only do that after `setopt | grep extendedglob` gives you a result.

**Comment by mcornella** [comment]
Friendly ping @TravisCarden

**Comment by TravisCarden** [comment]
Hey, @mcornella. I'm not actually quite sure how to act on what you wrote above. 😬

**Comment by mcornella** [comment]
Hi @TravisCarden, apologies for the delay.

If this is still happening, here's what you need to do so that I can debug it:

1. Run `zsh -lxv 2> >(tee ~/omz-debug.log &>/dev/null)` which will get you in a zsh shell while sending traces to `omz-debug.log`.
2. Keep using the shell as normally until `setopt | grep extendedglob` gives you a result.
3. At this point, run `exit` and review the instances of `extendedglob` in the trace file with `grep -i -E "extended.*glob" omz-debug.log`.

If you are unable to see the culprit, and have verified point 2, you can email the trace file to logs@ohmyz.sh: this email address will relay the email to me and @carlosala for us to review. _We are working on setting up a drop bucket with traceability for user logs that is secure and automated, but until then the email address is the best choice if you want security from prying eyes._

Let me know whichever you find or whether it's not relevant anymore. Thanks!

**Comment by TravisCarden** [comment]
> apologies for the delay.

As an open source project maintainer myself, I appreciate the cost of providing support, and I thank you for any time you give. 🙂

And I found the issue! In `~/.zlogin`:

```shell
[[ -s "$HOME/.rvm/scripts/rvm" ]] && source "$HOME/.rvm/scripts/rvm" # Load RVM into a shell session *as a function*
```

Why that should cause the problem I don't know. But since I scarcely ever use Ruby, I can just comment this out and worry about it another day. 😄

Thanks again for your help. 🙂

**Comment by mcornella** [comment]
That's great to hear! Thanks for following up with the answer.

I will then close the issue, we can reopen it if someone else has the same issue and it is not related to rvm.

Cheers!
